File: mad3004-introd-swift/C0711561_F2017_MAD3004_FinalProject/C0711561_F2017_MAD3004_FinalProject/Employee.swift
public class Employee: IPrintable {
    
    private var _name: String
    public var name: String {
        get {
            return _name
        }
        
        set {
            _name = newValue
        }
    }
    
    private var _age: Int
    public var age: Int {
        get {
            return _age
        }
        
        set {
            _age = newValue
        }
    }
    
    private var _v: Vehicle? // optional
    public var v: Vehicle? {
        get {
            return _v
        }
        
        set {
            _v = newValue
        }
    }
    
    init() {
        _name = ""
        _age = 0
        _v = nil
    }
    
    init (_ pName: String,_ pAge: Int) {
        _name = pName
        _age = pAge
        _v = nil
    }
    
    init (_ pName: String,_ pAge: Int, _ pV: Vehicle) {
        _name = pName
        _age = pAge
        _v = pV
    }
    
    func calcBirthYear() -> Int {
        return (2017 - self.age)
    }
    
    func getTypeOfEmployee() -> String {
        fatalError("This method must be overridden")
    }
    
    func calcEarnings() -> Double {
        fatalError("This method must be overridden")
    }
    
    func printMyData() -> String {
        let line = "\n-------------------------------------------"
        let personalInfo = "\nName: \(name)\nAge: \(age)\nYear of Birth: \(calcBirthYear())\n"
        
        var vehicleInfo = "\nEmployee has no vehicle registered"
        if let vehicle = v {
            vehicleInfo = "\nEmployee has a \(vehicle.getVehicleType())\n\(vehicle.printMyData())"
        }
        
        let employeeInfo = "\nEmployee is \(getTypeOfEmployee())"
        
        return line + personalInfo + vehicleInfo + employeeInfo
    }
}

File: mad3464-programming-java-i/C0711561_F2017_MAD3464_FinalProject/src/c0711561/f2017/mad3464/finalproject/employee/Employee.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package c0711561.f2017.mad3464.finalproject.employee;

import c0711561.f2017.mad3464.finalproject.IPrintable;
import c0711561.f2017.mad3464.finalproject.vehicle.Vehicle;
import java.time.LocalDate;

/**
 *
 * @author macstudent
 */
public abstract class Employee implements IPrintable {
    
    private final String name;
    private final int age;
    private Vehicle vehicle;

    public Employee(String name, int age) {
        this.name = name;
        this.age = age;
    }
    
    public Employee(String name, int age, Vehicle vehicle) {
        this.name = name;
        this.age = age;
        this.vehicle = vehicle;
    }
    
    public int calcBirthYear() {
        return LocalDate.now().getYear() - age;
    }
    
    public double calcEarnings() {
        return 1000;
    }
    
    public abstract String getTypeOfEmployee();
    
    @Override
    public String printMyData() {
        String line = "\n-------------------------------------------";
        
        String personalInfo = String.format("\nName: %s\nAge: %d\nYear of Birth: %d\n", name, age, calcBirthYear());
        
        String vehicleInfo;
        if (employeeHasVehicle()) {
            vehicleInfo = String.format("\nEmployee has a %s\n%s", vehicle.getVehicleType(), vehicle.printMyData());
        } else {
            vehicleInfo = "\nEmployee has no vehicle registered";
        }
        
        String employeeInfo = String.format("\nEmployee is %s", getTypeOfEmployee());
        
        return new StringBuilder().append(line).append(personalInfo).append(vehicleInfo).append(employeeInfo).toString();
    }

    private boolean employeeHasVehicle() {
        return vehicle != null;
    }
}
